### PR TITLE
Add API support to manage Telegram proxies.

### PR DESCRIPTION
**Pull request by davidgfnet**
This adds new endpoints: getProxies, addProxy, deleteProxy, enableProxy
and disableProxy.
The yaml doc file has been updated to hopefully correct information.

### CODE REVIEW

**Comment by MarcoBuster** [comment]
Hi, thank you for your PR! The documentation job is failing, please fix :)

**Comment by davidgfnet** [comment]
Hey mind taking another look at this?

**Comment by Cavallium** [comment]
@davidgfnet Please the next time don't force-push the pr, because we can't tell what you changed

**Review by Cavallium — APPROVED** [review]

File: telegram-bot-api/Client.cpp [hunk 1]
```diff
@@ -2686,6 +2691,56 @@ class Client::JsonMessagesArray : public Jsonable {
...
+      }
+      break;
+    case td_api::proxyTypeMtproto::ID:
+      {
+        auto ptype = static_cast<td_api::proxyTypeMtproto *>(proxy_->type_.get());
+        object("type", "mtproto");
+        object("secret", ptype->secret_);
+      }
+      break;
+    };
```
Comment (giuseppeM99) [inline telegram-bot-api/Client.cpp:2722]: you're missing https://core.telegram.org/tdlib/docs/classtd_1_1td__api_1_1proxy_type_http.html

File: telegram-bot-api/Client.cpp [hunk 2]
```diff
@@ -8638,6 +8733,66 @@ td::Status Client::process_get_memory_stats_query(PromisedQueryPtr &query) {
...
+  td_api::object_ptr<td_api::ProxyType> type;
+  td::to_lower_inplace(proxy_type);
+  if (proxy_type == "mtproto") {
+    TRY_RESULT(proxy_secret, get_required_string_arg(query.get(), "secret"));
+    type = td_api::make_object<td_api::proxyTypeMtproto>(proxy_secret.str());
+  } else if (proxy_type == "socks5") {
+    auto proxy_user = query->arg("username");
+    auto proxy_pass = query->arg("password");
+    type = td_api::make_object<td_api::proxyTypeSocks5>(proxy_user.str(), proxy_pass.str());
+  } else {
```
Comment (giuseppeM99) [inline telegram-bot-api/Client.cpp:8760]: you're missing https://core.telegram.org/tdlib/docs/classtd_1_1td__api_1_1proxy_type_http.html

File: telegram-bot-api/Client.cpp [hunk 3]
```diff
@@ -8638,6 +8733,66 @@ td::Status Client::process_get_memory_stats_query(PromisedQueryPtr &query) {
...
+    return Status::Error(400, PSLICE() << "Invalid proxy_id specified");
+  }
+
+  send_request(make_object<td_api::removeProxy>(pid),
+               std::make_unique<TdOnOkQueryCallback>(std::move(query)));
+  return Status::OK();
+}
+
+td::Status Client::process_enable_proxy_query(PromisedQueryPtr &query) {
+  int32 pid = get_integer_arg(query.get(), "proxy_id", 0, 0);
```
Comment (giuseppeM99) [inline telegram-bot-api/Client.cpp:8781]: are you sure that the proxy_id cannot be 0?

**Comment by davidgfnet** [comment]
Ok I did not force push it this time :)

**Review by giuseppeM99 — APPROVED** [review]

**Comment by luckydonald** [comment]
I have a few points I think we still need to address, and at least document:

- A section about proxy support listing the available methods in the README would be nice for those not looking at the changelog too closely.

- How can the configuration of that be turned off for the public instances we host?

**Comment by Cavallium** [comment]
> I have a few points I think we still need to address, and at least document:
> 
> * A section about proxy support listing the available methods in the README would be nice for those not looking at the changelog too closely.
> * How can the configuration of that be turned off for the public instances we host?

Open a PR with the changes or a new issue, so we can track this problem

**Pull request merged**